man/ptof.1
.\"
.\" $Id: ptof.1,v 1.2 2003/08/11 12:09:14 vflegel Exp $
.\" Copyright (c) 2003 SIB Swiss Institute of Bioinformatics <pftools@sib.swiss>
.\" Process this file with
.\" groff -man -Tascii <name>
.\" for ascii output or
.\" groff -man -Tps <name>
.\" for postscript output
.\"
.TH PTOF 1 "August 2003" "pftools 2.3" "pftools"
.\" ------------------------------------------------
.\" Name section
.\" ------------------------------------------------
.SH NAME
ptof \- convert a protein profile into a frame-search profile  
.\" ------------------------------------------------
.\" Synopsis section
.\" ------------------------------------------------
.SH SYNOPSIS
.TP 10
.B ptof
[
.B \-hlr
] [
.B \-B
.I init/term_score
] [
.B \-F
.I frameshift
] [
.B \-I
.I insert_multiplier
] [
.B \-X
.I stop_codon
] [
.B \-Y
.I intron_open
] [
.B \-Z
.I intron_extend
] [
.I protein_profile
|
.B \-
] [
.I parameters
]
.\" ------------------------------------------------
.\" Description section
.\" ------------------------------------------------
.SH DESCRIPTION
.B ptof
converts a protein profile (generated for instance by pftools
programs 
.BR pfmake (1), 
.BR gtop (1)
or
.BR htop (1))
into a so-called
.RI \(dq frame-search\ profile \(dq.
A frame-search profile is used to search an
.RI \(dq interleaved\ frame-translated \(dq
DNA sequence (generated by pftools program
.BR 2ft (1))
for occurrences of a protein sequence motif.
An
.RI \(dq interleaved\ frame-translated \(dq
DNA sequence is an amino acid sequence corresponding to the 
.I N-2
overlapping codons of a DNAsequence of length
.IR N .
Note that in such a sequence, the character
.RB ' O '
is used to represent stop codons.
.LP
The conversion procedure works as follows: The protein profile is 
expanded in length by a factor of three to accommodate three translated
codons per original match position.
Two dummy match positions are placed between two consecutive significant match positions
imported from the original profile. 
The original insert positions are placed between pairs of adjacent dummy match positions.
The initiation, termination, and transition scores of the original insert
positions are left unchanged; the
insert extension scores are divided by a factor of 3, or by the value of the command-line
option 
.BR \-I .  
The two insert positions flanking the significant match positions 
serve to accommodate frame-shift errors and introns, respectively.
The frame-shift insert position allows free insertion opening combined
with a high insert extension penalty (command-line option 
.BR \-F ) 
whereas the intron insertion position has high opening but low extension penalties
(command line options 
.B \-Y
and
.BR \-Z ).
The deletion opening and closing penalties next to the significant 
match positions are set to values that ensure that the total cost of a single-base
deletion is the same as the cost of  
a single base-insertion at a frame-shift insert position.
Furthermore, the alphabet of the original profile is extended by the stop codon symbol
.RB ' O ' 
which is assigned a constant negative value (command-line option
.BR \-X )
at significant match positions, and zero at dummy match positions. At insert positions,
it is set to the average of the other insert extension scores.
.\" ------------------------------------------------
.\" Options section
.\" ------------------------------------------------
.SH OPTIONS
.\" --- protein_profile ---
.TP
.I protein_profile
Input protein profile.
.br
The protein profile contained in this file will be converted into a frame-search profile.
If the filename is replaced by a
.RB ' \- ',
.B ptof
will read the profile from
.BR stdin .
.\" --- h ---
.TP
.B \-h
Display usage help text.
.\" --- l ---
.TP
.B \-l
Remove output line length limit. Individual lines of the output profile
can exceed a length of 132 characters, removing the need to wrap them over several lines. 
.\" --- r ---
.TP
.B \-r
Frame-search parameters are given in normalized score units. This option
will only be considered if a linear normalization function with
priority over all other normalization functions is specified 
in the profile. 
In this case, the frame-search scores specified on the command line will be 
divided by the slope
.RI ( R2
parameter) of the normalization function. 
This option is particularly useful for profiles which are 
already scaled in units that can be interpreted as \(miLog(P)-values, 
.I e.g.
bits.  
.\" --- B ---
.TP
.BI \-B\  init/term_score
Minimal initiation and termination score.
.br
All internal and external initiation and termination scores will be set to this 
value if the corresponding value in the original profile is lower than 
this value.
This parameter is used to impose a more
local alignment behavior on the frame-search profile in order to
deal with discontinuities in DNA sequences (long introns,
alternative splicing, chimeric clones, 
.IR etc. )
.br 
Default: \-50 (\-0.5 with option
.BR \-r )
.\" --- F ---
.TP
.BI \-F\  frameshift
Frame-shift error penalty.
.br
Default: \-100 (\-1.0 with option
.BR \-r ) 
.\" --- I ---
.TP
.BI \-I\  insert_multiplier
Insert score multiplier.
.br
The values of the original
insert extension scores will be multiplied by this
factor in order to compensate for the fact that 
a single amino acid corresponds to three overlapping
codon positions in the target sequence.
.br
Default: 1/3
.\" --- X ---
.TP
.BI \-X\  stop_codon
Stop codon penalty.
.br
Default: \-100 (\-1.0 with option
.BR -r )
.\" --- Y ---
.TP
.BI \-Y\  intron_open
Intron opening penalty.
.br
Default: \-300 (\-3.0 with option
.BR \-r ) 
.\" --- Z ---
.TP
.BI \-Z\  intron_extend
Intron extension penalty.
.br
Default: \-1 (\-0.01 with option
.BR \-r )
.\" ------------------------------------------------
.\" Parameters section
.\" ------------------------------------------------
.SH PARAMETERS
.TP
Note:
for backwards compatibility, release 2.3 of the
.B pftools
package will parse the version 2.2 style parameters, but these are
.I deprecated
and the corresponding option (refer to the
.I options
section) should be used instead.
.TP
B=#
Minimal initiation and termination score.
.br
Use option
.B \-B
instead.
.TP
F=#
Frame-shift error penalty.
.br
Use option
.B \-F
instead.
.TP
I=#
Insert score multiplier.
.br
Use option
.B \-I
instead.
.TP
X=#
Stop codon penalty.
.br
Use option
.B \-X
instead.
.TP
Y=#
Intron opening penalty.
.br
Use option
.B \-Y
instead.
.TP
Z=#
Intron extension penalty.
.br
Use option
.B \-Z
instead.
.\" ------------------------------------------------
.\" Examples section
.\" ------------------------------------------------
.SH EXAMPLES
.TP
(1)
.B ptof
\-r \-F -1.2 \-I 0.6 \-X -1.5 \-B -0.5 sh3.prf > sh3.fsp
.br
.B 2ft \-
< R76849.seq | 
.B pfsearch 
\-fy \-C 5.0 sh3.fsp \- 
.IP
The protein domain profile in 
.RI ' sh3.prf '
is first converted into a frame-search profile 
.RI ' sh3.fsp '.
Then both strands of the Fasta-formatted EST sequence in
.RI ' R76849.seq '
(GenBank/EMBL-accession: R76849) are converted into interleaved frame-translated
protein sequences and searched for SH3 domains with the frame-search profile 
generated in the preceding step.
.IP
The output may be compared to the result of a more conventional
search strategy using a protein profile in conjunction with a six-frame
translation of the same DNA sequence:
.IP
.B 6ft \-
< R76849.seq | 
.B pfsearch 
\-fy \-C 5.0 sh3.prf \-
.\" ------------------------------------------------
.\" Exit code section
.\" ------------------------------------------------
.SH EXIT CODE
.LP
On successful completion of its task,
.B ptof
will return an exit code of 0. If an error occurs, a diagnostic message will be
output on standard error and the exit code will be different from 0. When conflicting
options where passed to the program but the task could nevertheless be completed, warnings
will be issued on standard error.
.\" ------------------------------------------------
.\" See also section
.\" ------------------------------------------------
.SH "SEE ALSO"
.BR pfscan (1),
.BR pfsearch (1),
.BR 2ft (1),
.BR 6ft (1)
.\" ------------------------------------------------
.\" Author section
.\" ------------------------------------------------
.SH AUTHOR
The
.B pftools
package was developed by Philipp Bucher.
.br
Any comments or suggestions should be addressed to <pftools@sib.swiss>.


src/Fortran/ptof.f
*       Program ptof 
*----------------------------------------------------------------------*     
* $Id: ptof.f,v 2.12 2003/12/01 13:33:04 vflegel Exp $
*----------------------------------------------------------------------*     
*       Function: converts a protein profile into a framesearch profile
*       Author:   Philipp Bucher
*       Contact:  pftools@sib.swiss
*       Version:  File under developpment for release 2.3
*----------------------------------------------------------------------*     
* DATA
*----------------------------------------------------------------------*     

* array dimensions and I/O units

      Include          'ardim.f' 

      Parameter        (NOUT=   6)    

      Parameter        (NPRF=  11)
      Parameter        (NNUL=  12)

* profile 

      Character*512     FPRF

      Include          'psdat.f'
      Include          'gsdat.f'
      Include          'djdat.f'
      Include          'nodat.f'
      Include          'codat.f'
      Include          'pfdat.f'
      Include          'dfdat.f'

      Include          'sterr.f'


* function return types

      Integer           Xblnk
      External          Xblnk

* parameters and options

      Logical           OPTR
      Logical           LLLT

      Real              RB
      Real              RF
      Real              RI
      Real              RX 
      Real              RY 
      Real              RZ 

      Integer           INBP
      Logical           LEOF
      Logical           LRNM

* initialization of controlled vocabularies

      Include          'cvini.f' 

*----------------------------------------------------------------------*     
* INPUT SECTION 
*----------------------------------------------------------------------*     

      IRC=0
      FPRF='stdout'
      LRNM=.FALSE.
      INBP=0

* command line arguments
      
      Call Repar(FPRF,OPTR,LLLT,RB,RF,RI,RX,RY,RZ,IRC)
      If(IRC.NE.0) then
         Write(NERR,'(/,
     *      ''ptof 2.3 revision 5.d'',//
     *      ''Usage: ptof [ -hlrBFIXYZ ] [ profile | - ] '',
     *      ''[ parameters ]'',//
     *      )')
         Write(NERR,'(
     *      ''   options:'',/,
     *      ''    -h: print usage help text.'',/
     *      ''    -l: do not impose limit on line length.'',/
     *      ''    -r: parameters given as normalized score units.''
     *      )')
         Write(NERR,'(
     *      ''    -B<value>:'',/
     *      ''        minimal initiation/termination score (default: '',
     *      ''-50 or -0.5 with option -r).'',/
     *      ''    -F<value>:'',/
     *      ''        frameshift error penalty (default: '',
     *      ''-100 or -1.0 with option -r).'',/
     *      ''    -I<value>:'',/
     *      ''        insert score multiplier (default: 1/3).'',/
     *      ''    -X<value>:'',/
     *      ''        stop codon penalty (default: '',
     *      ''-100 or -1.0 with option -r).'',/
     *      ''    -Y<value>:'',/
     *      ''        intron opening penalty (default: '',
     *      ''-300 or -3.0 with option -r).'',/
     *      ''    -Z<value>:'',/
     *      ''        intron extension penalty (default: '',
     *      ''-1 or -0.01 with option -r).'',/
     *      )')
         Write(NERR,'(
     *      ''   valid (but deprecated) parameters are:'',/,
     *      ''    [B=min-beg-end-score]   use option -B instead'',/
     *      ''    [F=deletion-frameshift] use option -F instead'',/
     *      ''    [I=insert-score-mult]   use option -I instead'',/
     *      ''    [X=stop-codon-score]    use option -X instead'',/
     *      ''    [Y=intron-open-penalty] use option -Y instead'',/
     *      ''    [Z=intron-ext-penalty]  use option -Z instead'',/
     *      )')
         Call Exit(IRC)
      End if

* read profile
      
      If(FPRF.EQ.'-') then
         MPRF=5
      Else
         MPRF=NPRF
      End if

      Call REPRF
     *   (MPRF,FPRF,
     *   CPID,CPAC,CPDT,CPDE,LHDR,CHDR,LFTR,CFTR,NABC,CABC,LPRF,LPCI,
     *   BLOG,FABC,P0,
     *   CDIS,JDIP,MDIS,NDIP,
     *   CNOR,JNOP,JNOR,MNOR,NNOR,NNPR,CNTX,RNOP,
     *   JCUT,MCLE,CCUT,ICUT,JCNM,RCUT,MCUT,
     *   IDMP,CHIP,IIPP,CHMP,IMPP,CHIL,IIPL,ILIP,
     *   CHID,IIPD,CHMD,IMPD,
     *   INBP,LEOF,.FALSE.,IRC)

      If(IRC.NE.0) go to 100

* expanded profile size check

      If(LPRF*3.GT.IDMP) go to 901
      If(NABC.LE.4) go to 920

* add command-line to footer lines

 10   Continue
      LFTR=LFTR+1
      If(LFTR.GT.1024) LFTR=1024
      Do I1=LFTR,2,-1
         CFTR(I1)=CFTR(I1-1)
      End do 

      CFTR(1)='CC   /GENERATED_BY="'
      Call Recmd(CFTR(1)(21:130))
      IC=Lblnk(CFTR(1))
      CFTR(1)(IC+1:)='";'

*----------------------------------------------------------------------*     
* CONVERSION SECTION
*----------------------------------------------------------------------*     

* Adjust frame search parameters to normalization function      

      If(OPTR) then
         R2=0.01 
         K1=1
         J1=NNPR( 1)
         Do I1=2,JNOR
            If(NNPR(I1).LT.J1) then
               K1=I1
               JN=NNPR(I1) 
            End if
         End do
         If(MNOR(K1).EQ.1) R2=RNOP(2,K1)
         
         RB=RB / R2
         RF=RF / R2
         RX=RX / R2
         RY=RY / R2
         RZ=RZ / R2
      End if 

* Expand alphabet

      NABC=NABC+1
      CABC(NABC)='O'
      NX=NINT(RX)
      J2=0
      Do I2=1,NABC-1
         J2=J2+IIPD(I2)
      End do
      IIPD(NABC)=J2/(NABC-1)
      J2=0
      Do I2=1,NABC-1
         J2=J2+IIPP(I2,0)
      End do
      IIPP(NABC,0)=J2/(NABC-1)
      Do I1=1,LPRF
         IMPP(NABC,I1)=NX
         J2=0
         Do I2=1,NABC-1
            J2=J2+IIPP(I2,I1)
         End do
         IIPP(NABC,I1)=J2/(NABC-1)
      End do 

* - default begin, end score

      IIPD(B0)=MAX(IIPD(B0),NINT(RB))
      IIPD(B1)=MAX(IIPD(B1),NINT(RB))
      IIPD(E0)=MAX(IIPD(E0),NINT(RB))
      IIPD(E1)=MAX(IIPD(E1),NINT(RB))

* - eliminate II transitions

      Do I1=0,LPRF
         If(IIPP(BI,I1).NE.NLOW) IIPP(BI,I1)=IIPP(BI,I1)-IIPP(II,I1)
         If(IIPP(MI,I1).NE.NLOW) IIPP(MI,I1)=IIPP(MI,I1)-IIPP(II,I1)
         If(IIPP(DI,I1).NE.NLOW) IIPP(DI,I1)=IIPP(DI,I1)-IIPP(II,I1)
         Do I2=0,26
            If(IIPP(I2,I1).NE.NLOW)
     *         IIPP(I2,I1)=IIPP(I2,I1)+IIPP(II,I1)
         End do
         IIPP(II,I1)=0
      End do       

* Expand profile

* -define architecture of profile according to topology 

      If(LPCI) then 
         NBB=1
         NB=2
         NE=3*LPRF-1
         NEE=3*LPRF
      Else 
         NBB=1
         NB=1
         NE=3*LPRF-2
         NEE=3*LPRF-2
      End if       

* - move last insert position to new end of profile

      Do I1=0,46
         IIPP(I1,NEE)=IIPP(I1,LPRF)
      End do 
      CHIP(NEE)=CHIP(LPRF)

* - move other positions 

      K1=NE
      Do I1=LPRF,1,-1
         Do I2=0,27
            IMPP(I2,K1)=IMPP(I2,I1) 
         End do
         CHMP(K1)=CHMP(I1)
         K1=K1-3
      End do 

      K1=NE-2
      Do I1=LPRF-1,1,-1
         Do I2= 0,26
            IIPP(I2,K1)=NINT(IIPP(I2,I1)*RI)
         End do
         Do I2=27,46
            IIPP(I2,K1)=IIPP(I2,I1)
         End do
         CHIP(K1)=CHIP(I1)
         K1=K1-3
      End do 

* - fill in m-1, m+1 positions

      Do I1=NB+1,NEE,3
         Do I2=0,27
            IMPP(I2,I1)=0
         End do
         CHMP(I1)='>'
      End do

      J1=NB-1
      If(J1.LE.0) J1=J1+3
      Do I1=J1,NEE,3
         Do I2=0,27
            IMPP(I2,I1)=0
         End do
         CHMP(I1)='<'
      End do
      
*- fill i-1 position 
      
      Do I2= 0,26
         IIPP(I2,NB)=0
      End do
      Do I2=27,46
         IIPP(I2,NB)=NLOW
      End do
      IIPP(MM,NB)=0
      IIPP(DD,NB)=0
      IIPP(MI,NB)=NINT(RF/2)
      IIPP(II,NB)=NINT(RF)
      IIPP(IM,NB)=NINT(RF/2)
      CHIP(NB)='*'
      Do I1=NB+3,NEE-1,3
         Do I2= 0,47
            IIPP(I2,I1)=IIPP(I2,NB)
         End do
         CHIP(I1)='*'
      End do 

* - fill i+1 position 
      
      J1=3
      If(LPCI) J1=1
      Do I2= 0,26
         IIPP(I2,J1)=0
      End do
      Do I2=27,46
         IIPP(I2,J1)=NLOW
      End do
      IIPP(MM,J1)=0
      IIPP(DD,J1)=0
      IIPP(MI,J1)=NINT(RY/2)
      IIPP(II,J1)=MIN(-1,NINT(RZ))
      IIPP(IM,J1)=NINT(RY/2)
      CHIP(J1)=':'
      Do I1=J1+3,NEE-1,3
         Do I2=0,46
            IIPP(I2,I1)=IIPP(I2,J1)
         End do
         CHIP(I1)=':'
      End do 

* frameshift deletetion scores: 

      J1=2   
      If(LPCI) J1=1
      K1=3
      If(LPCI) K1=1
      Do I1=J1,LPRF-1
         IIPP(MD,K1)=(NINT(RF-IMPP( D,K1+1))/2)
         IIPP(DM,K1+1)=NINT(RF-IIPP(MD,K1)-IMPP( D,K1+1))
         K1=K1+3
      End do 

* begin and end scores

      If(.NOT.LPCI) then
         Do I1=2,NEE,3
            IIPP(B0,I1-1)=IIPP(B0,I1)
            IIPP(B1,I1-1)=IIPP(B1,I1) 
            IIPP(BM,I1-1)=IIPP(BM,I1) 
            IIPP(BI,I1-1)=IIPP(BI,I1) 
            IIPP(BD,I1-1)=IIPP(BD,I1) 
            IIPP(E0,I1-1)=IIPP(E0,I1)
            IIPP(E1,I1-1)=IIPP(E1,I1)
            IIPP(ME,I1-1)=IIPP(ME,I1) 
            IIPP(IE,I1-1)=IIPP(IE,I1) 
            IIPP(DE,I1-1)=IIPP(DE,I1) 
            IIPP(B0,I1+1)=IIPP(B0,I1)
            IIPP(B1,I1+1)=IIPP(B1,I1)
            IIPP(BM,I1+1)=IIPP(BM,I1) 
            IIPP(BI,I1+1)=IIPP(BI,I1) 
            IIPP(BD,I1+1)=IIPP(BD,I1) 
            IIPP(E0,I1+1)=IIPP(E0,I1)
            IIPP(E1,I1+1)=IIPP(E1,I1)
            IIPP(ME,I1+1)=IIPP(ME,I1) 
            IIPP(IE,I1+1)=IIPP(IE,I1) 
            IIPP(DE,I1+1)=IIPP(DE,I1) 
         End do
      Else
         Do I1=2,NEE,3
            IIPP(B0,I1-1)=IIPP(B0,I1-2)
            IIPP(B1,I1-1)=IIPP(B1,I1-2) 
            IIPP(BM,I1-1)=IIPP(BM,I1-2) 
            IIPP(BI,I1-1)=IIPP(BI,I1-2) 
            IIPP(BD,I1-1)=IIPP(BD,I1-2) 
            IIPP(E0,I1-1)=IIPP(E0,I1-2)
            IIPP(E1,I1-1)=IIPP(E1,I1-2)
            IIPP(ME,I1-1)=IIPP(ME,I1-2) 
            IIPP(IE,I1-1)=IIPP(IE,I1-2) 
            IIPP(DE,I1-1)=IIPP(DE,I1-2) 
            IIPP(B0,I1  )=IIPP(B0,I1+1)
            IIPP(B1,I1  )=IIPP(B1,I1+1)
            IIPP(BM,I1  )=IIPP(BM,I1+1) 
            IIPP(BI,I1  )=IIPP(BI,I1+1) 
            IIPP(BD,I1  )=IIPP(BD,I1+1) 
            IIPP(E0,I1  )=IIPP(E0,I1+1)
            IIPP(E1,I1  )=IIPP(E1,I1+1)
            IIPP(ME,I1  )=IIPP(ME,I1+1) 
            IIPP(IE,I1  )=IIPP(IE,I1+1) 
            IIPP(DE,I1  )=IIPP(DE,I1+1) 
         End do
      End if 

      Do I1=NBB,NEE-1
         IIPP(B0,I1)=MAX(IIPD(B0),IIPP(B0,I1)) 
         IIPP(B1,I1)=MAX(IIPD(B1),IIPP(B1,I1)) 
         IIPP(E0,I1)=MAX(IIPD(E0),IIPP(E0,I1)) 
         IIPP(E1,I1)=MAX(IIPD(E1),IIPP(E1,I1)) 
      End do

* disjointness definition

      If(MDIS.EQ.2) then
         NDIP(1)=NDIP(1)*3-2
         NDIP(2)=NDIP(2)*3-2
      End if

* adjust length

      LPRF=NEE

*----------------------------------------------------------------------*     
* OUTPUT SECTION
*----------------------------------------------------------------------*     

* write profile

      Call WRPRF
     *   (NOUT,LLLT,LRNM,
     *   CPID,CPAC,CPDT,CPDE,LHDR,CHDR,LFTR,CFTR,NABC,CABC,LPRF,LPCI,
     *   CDIS,JDIP,MDIS,NDIP,
     *   CNOR,JNOP,JNOR,MNOR,NNOR,NNPR,CNTX,RNOP,
     *   JCUT,MCLE,CCUT,ICUT,JCNM,RCUT,MCUT,
     *   IDMP,CHIP,IIPP,CHMP,IMPP,
     *   BLOG,FABC,P0,
     *   CHID,IIPD,CHMD,IMPD,
     *   IRC)

 100  Call Exit(IRC)

* errors

 901  Write(NERR,*) 'Error: Expanded profile length exceeds buffer ',
     *   'size (',IDMP,').'
      IRC=1
      Go to 100

* warnings

 920  Write(NERR,*) 'Warning: Profile seems to be DNA. Check input '//
     *   'file.'
      Go to 10

      End
*----------------------------------------------------------------------*
      Subroutine Repar
     *   (FPRF,OPTR,LLLT,RB,RF,RI,RX,RY,RZ,IRC)

      Character*(*)     FPRF
      Character*512     CARG 

      Logical           OPTR
      Logical           LLLT
      Real              RB
      Real              RF
      Real              RI
      Real              RX
      Real              RY
      Real              RZ

      IRC=0
      
* initializations

      FPRF='-'

      OPTR=.FALSE.
      LLLT=.TRUE.
      RB=-50
      RF=-100
      RI=1/3.0
      RX=-100
      RY=-300
      RZ=-1
      
      N1=Iargc()

      K1=0
      I2=1
      Do I1=1,N1
         Call GetArg(I2,CARG)
         If     (CARG(1:1).EQ.'-'.AND.CARG(2:2).NE.' ') then
            If(Index(CARG,'h').NE.0) go to 900
            If(Index(CARG,'l').NE.0) LLLT=.FALSE.
            If(Index(CARG,'r').NE.0) then
               OPTR=.TRUE.
               RB=-0.5
               RF=-1.0
               RX=-1.0
               RY=-3.0
               RZ=-0.01
            End if
            If(Index(CARG,'B').NE.0) then
               If(CARG(3:3).NE.' ') then
                  Read(CARG(3:),*,Err=900) RB
               Else
                  I2=I2+1
                  Call GetArg(I2,CARG)
                  Read(CARG,*,Err=900) RB
               End if
            End if
            If(Index(CARG,'F').NE.0) then
               If(CARG(3:3).NE.' ') then
                  Read(CARG(3:),*,Err=900) RF
               Else
                  I2=I2+1
                  Call GetArg(I2,CARG)
                  Read(CARG,*,Err=900) RF
               End if
            End if
            If(Index(CARG,'I').NE.0) then
               If(CARG(3:3).NE.' ') then
                  Read(CARG(3:),*,Err=900) RI
               Else
                  I2=I2+1
                  Call GetArg(I2,CARG)
                  Read(CARG,*,Err=900) RI
               End if
            End if
            If(Index(CARG,'X').NE.0) then
               If(CARG(3:3).NE.' ') then
                  Read(CARG(3:),*,Err=900) RX
               Else
                  I2=I2+1
                  Call GetArg(I2,CARG)
                  Read(CARG,*,Err=900) RX
               End if
            End if
            If(Index(CARG,'Y').NE.0) then
               If(CARG(3:3).NE.' ') then
                  Read(CARG(3:),*,Err=900) RY
               Else
                  I2=I2+1
                  Call GetArg(I2,CARG)
                  Read(CARG,*,Err=900) RY
               End if
            End if
            If(Index(CARG,'Z').NE.0) then
               If(CARG(3:3).NE.' ') then
                  Read(CARG(3:),*,Err=900) RZ
               Else
                  I2=I2+1
                  Call GetArg(I2,CARG)
                  Read(CARG,*,Err=900) RZ
               End if
            End if
         Else if(CARG(1:2).EQ.'B=') then
            Read(CARG(3:),*,Err=900) RB 
         Else if(CARG(1:2).EQ.'F=') then
            Read(CARG(3:),*,Err=900) RF 
         Else if(CARG(1:2).EQ.'I=') then
            Read(CARG(3:),*,Err=900) RI
         Else if(CARG(1:2).EQ.'X=') then
            Read(CARG(3:),*,Err=900) RX
         Else if(CARG(1:2).EQ.'Y=') then
            Read(CARG(3:),*,Err=900) RY
         Else if(CARG(1:2).EQ.'Z=') then
            Read(CARG(3:),*,Err=900) RZ
         Else if(K1.LE.0) then
            K1=K1+1
            If     (K1.EQ.1) then
               FPRF=CARG
            End if
         Else
            Go to 900
         End if
         I2=I2+1
         If(I2.GT.N1) Go to 20
      End do

 20   Continue
      
 100  Return
 900  IRC=-1   
      Go to 100
      End
*----------------------------------------------------------------------*
      Include          'reprf.f'
      Include          'recmd.f'
      Include          'wrprf.f'
      Include          'lblnk.f' 
      Include          'Xblnk.f'
